### PR TITLE
Authorization

### PR DESCRIPTION
**Pull request by bartoval**
Note:
- Red hat sso apis
https://sso-sso.vabar-vpc-cluster-153f1de160110098c1928a6c05e19444-0000.eu-gb.containers.appdomain.cloud/auth/realms/adminrealm/.well-known/openid-configuration

token expire after 2 minutes (test)

need to improve:
- firefox does not display the 'browser message page' after login (black page)
- use an other github account
- run red had sso in an other openshift space

### CODE REVIEW

**Comment by thoraxe** [comment]
@bartoval this is awesome. It needs a few enhancements:
1) The auth flow needs to be optional for people to do testing without it
2) Can you provide some documentation on how to run keycloak in a container locally for using/testing this feature?

**Comment by bartoval** [comment]
> @bartoval this is awesome. It needs a few enhancements:
> 
> 1. The auth flow needs to be optional for people to do testing without it
> 2. Can you provide some documentation on how to run keycloak in a container locally for using/testing this feature?

Thanks for these feedbacks. 
I updated the PR

ps. Maybe it worth it create a Resources/config.cfg.template file  and ban from .githignore the config.cfg

**Comment by thoraxe** [comment]
It looks like you might have some merge issues. I commented on them in the files themselves.

You definitely have file spacing issues. We are using spaces (not tabs) and an indent of 2. It's making some things look like changes that are not changes.

File: Scenes/MainScenes/Game.cs [hunk 1]
```diff
@@ -656,6 +673,12 @@ public SpaceMissile CreateMissileForUUID(GameEvent egeb)
     AddChild(missileInstance);
     //missileInstance.GetNode<AudioStreamPlayer>("FireSound").Play();
 
+    missileObjects.Add(egeb.Uuid, missileInstance);
+    missileInstance.GlobalPosition = new Vector2(egeb.PositionX, egeb.PositionY);
+    missileInstance.RotationDegrees = egeb.Angle;
+    _serilogger.Debug("Game.cs: Adding missile to scene tree");
+    AddChild(missileInstance);
+
```
Comment (thoraxe) [inline Scenes/MainScenes/Game.cs:681]: Is this a bd merge? It looks like it's repeating code from the previous lines.

File: Scenes/MainScenes/Game.cs [hunk 2]
```diff
@@ -703,6 +726,9 @@ void CreateLocalMissileForUUID(string uuid)
     // players own missiles, inversely
     missileInstance.MyPlayer.MyMissile = missileInstance;
 
+    // players own missiles, inversely
+    missileInstance.MyPlayer.MyMissile = missileInstance;
```
Comment (thoraxe) [inline Scenes/MainScenes/Game.cs:730]: again this looks like a bad merge

**Comment by thoraxe** [comment]
I'm not entirely clear on the "flow" of the code here. It looks like the authorization scene is added and its ready function calls the auth. I'm not sure why you're turning the `process` on and off for that node. Can you maybe add more code comments that are describing what things are doing and why?

Also, it's not clear why you are using a lot of async/await, so some commentary there would additionally be super helpful.

This is looking really good!

**Comment by thoraxe** [comment]
> ps. Maybe it's worth it creates a Resources/config.cfg.template file and ban from .githignore the config.cfg (same for the .vcode folder)

You can totally do this. This is a great idea.

**Comment by bartoval** [comment]
> It looks like you might have some merge issues. I commented on them in the files themselves.
> 
> You definitely have file spacing issues. We are using spaces (not tabs) and an indent of 2. It's making some things look like changes that are not changes.

I added a .editorconfig file. It should force vscode or other IDE's to follow these roles

**Comment by bartoval** [comment]
> > ps. Maybe it's worth it creates a Resources/config.cfg.template file and ban from .githignore the config.cfg (same for the .vcode folder)
> 
> You can totally do this. This is a great idea.

I added .template and removed the files without suffix. Probably, I should highlight in the Readme that now we need to remove the suffix and fill properly the file to work with the client.

**Comment by bartoval** [comment]
> I'm not entirely clear on the "flow" of the code here. It looks like the authorization scene is added and its ready function calls the auth. I'm not sure why you're turning the `process` on and off for that node. Can you maybe add more code comments that are describing what things are doing and why?
> 
> Also, it's not clear why you are using a lot of async/await, so some commentary there would additionally be super helpful.
> 
> This is looking really good!

Regarding the process on/off, I removed it. Now, I don't need it anymore. I used that during my first implementation, because I needed to catch the response from the auth callback, only in a particular window of time.

Regarding the flow I added some comments to the authorize function. 
I use async/awaits because I have 3 or 4 https requests to implement.
Alternatively, I can make requests and listen the responses to other events but I don't like this solution because it is verbose, less compact and hard to follow.

File: README.md [hunk 3]
```diff
@@ -42,3 +42,50 @@ This repo is the player game client.
...
+
+#### create github auth app
+- Go in your github account and click settings
+
+- Go in Developer settings -> Oauth app then click on the button "new Oauth app"
+
+- Here you need to set the "Authorization callback URL" with the redirect uri from keycloak (see section below)
+
+-  "Homepage URL" is mandatory but you can use any words
+#### Setup github
```
Comment (thoraxe) [inline README.md:73]: Is this done inside Keycloak?

File: Scenes/Authorization.cs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+        GD.Print("stop server");
+        connection.DisconnectFromHost();
+        redirectServer.Stop();
+      }
+
+    }
+  }
+  public async void authorize()
+  {
+    GD.Print("Authorizing");
```
Comment (thoraxe) [inline Scenes/Authorization.cs:78]: Please use the `_serilogger` for logging like we are doing in the other files.

File: Scenes/Authorization.cs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    string[] header = { "Content-Type:application/x-www-form-urlencoded" };
+
+    string[] bodyPart = {
+  String.Format("client_id={0}", clientID),
+  String.Format("client_secret={0}", clientSecret),
+  String.Format("token={0}", token),
+  "token_type_hint=access_token"
+  };
```
Comment (thoraxe) [inline Scenes/Authorization.cs:180]: I think the formatting/indentation on these lines is slightly off.

File: Scenes/Authorization.cs [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+
+    string[] bodyPart = {
+  String.Format("code={0}", authCode),
+  String.Format("client_id={0}", clientID),
+  String.Format("client_secret={0}", clientSecret),
+  String.Format("redirect_uri={0}", redirectUri),
+  "grant_type=authorization_code",
+  "scope=openId"
+  };
```
Comment (thoraxe) [inline Scenes/Authorization.cs:141]: Indentation?

File: Scenes/Authorization.cs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+
+    OS.ShellOpen(url);
+  }
+
+  private async Task<bool> getTokenFromAuthCode(string authCode)
+  {
+    GD.Print("get Token from AuthCode");
+    string[] header ={
+  "Content-Type:application/x-www-form-urlencoded"
+  };
```
Comment (thoraxe) [inline Scenes/Authorization.cs:131]: Indentation

File: Scenes/Authorization.cs [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+  {
+    GD.Print("call login - ask auth code");
+    redirectServer.Listen((ushort)PORT, HOST);
+
+    string[] bodyPart = {
+  String.Format("client_id={0}", clientID),
+  String.Format("redirect_uri={0}", redirectUri),
+  "response_type=code",
+  "scope=openid"
+  };
```
Comment (thoraxe) [inline Scenes/Authorization.cs:119]: Indentation

File: Scenes/Authorization.cs [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+
+    string url = String.Format("{0}?{1}", authServer, String.Join("&", bodyPart));
+
+    OS.ShellOpen(url);
+  }
+
+  private async Task<bool> getTokenFromAuthCode(string authCode)
```
Comment (thoraxe) [inline Scenes/Authorization.cs:126]: Why does this return a `Task<bool>` and not just `bool` ?

File: Scenes/Authorization.cs [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    string[] header = { "Content-Type:application/x-www-form-urlencoded" };
+
+    string[] bodyPart = {
+  String.Format("client_id={0}", clientID),
+  String.Format("client_secret={0}", clientSecret),
+  String.Format("refresh_token={0}", refreshToken),
+  "grant_type=refresh_token",
+  };
```
Comment (thoraxe) [inline Scenes/Authorization.cs:219]: Indentation

File: Scenes/Authorization.cs [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+    {
+      var tokens = new Dictionary();
+
+      tokens["token"] = token;
+      tokens["refreshToken"] = refreshToken;
+
+      file.StoreVar(tokens);
+      file.Close();
+
+      GD.Print("Token saved successfully");
```
Comment (thoraxe) [inline Scenes/Authorization.cs:257]: Use the logger

File: Scenes/LoginScreen.cs [hunk 12]
```diff
@@ -9,38 +9,46 @@ public class LoginScreen : Control
...
   LineEdit textField;
 
+	[Signal] public delegate void retryAuthorization();
+
   public override void _Ready()
   {
-    MyGame = GetNode<Game>("/root/Game");
-    _serilogger = MyGame._serilogger;
+	MyGame = GetNode<Game>("/root/Game");
+	_serilogger = MyGame._serilogger;
```
Comment (thoraxe) [inline Scenes/LoginScreen.cs:17]: Indentation

File: Scenes/LoginScreen.cs [hunk 13]
```diff
@@ -9,38 +9,46 @@ public class LoginScreen : Control
...
 
     textField = this.GetNode<LineEdit>("VBoxContainer/HBoxContainer/NameLineEdit");
     textField.GrabFocus();
 
-    // TODO: need to interrogate server for the initial defaults for things like
-    // missile speed
+	// TODO: need to interrogate server for the initial defaults for things like
+	// missile speed
```
Comment (thoraxe) [inline Scenes/LoginScreen.cs:23]: Indentation

File: Scenes/LoginScreen.cs [hunk 14]
```diff
@@ -9,38 +9,46 @@ public class LoginScreen : Control
...
+	  // TODO: alert errors or something 
+	}
+	else
+	{
+	  // since we successfully joined the game, we can remove this node
+	  // which is the login screen. removing the login screen "displays"
+	  // the main game window
+	  QueueFree();
+	  MyGame.initializeGameUI();
+	}
```
Comment (thoraxe) [inline Scenes/LoginScreen.cs:45]: Indentation

File: Scenes/MainScenes/Game.cs [hunk 15]
```diff
@@ -164,10 +164,27 @@ public override void _Ready()
...
+ var clientConfig = new ConfigFile();
+
+	Godot.Error err = clientConfig.Load("res://Resources/client.cfg");	  
+	var  activateAuthDev = (Boolean)clientConfig.GetValue("auth", "activate_auth_dev");
+
+	if (OS.IsDebugBuild() && activateAuthDev == false)
+	{ 
+		this._on_go_to_game(true);
+		return;
+	}
```
Comment (thoraxe) [inline Scenes/MainScenes/Game.cs:180]: Indentation

Comment (thoraxe) [reply Scenes/MainScenes/Game.cs:180]: Also, there is already a `LoadConfig()` method -- shouldn't this config loading be part of that method?

**Review by thoraxe — CHANGES_REQUESTED** [review]

File: Scenes/Authorization.cs (hunk 9 above)
Comment (bartoval) [reply Scenes/Authorization.cs:133]: Because this is an async function.
if I run the code using bool as return type, Godot fail and say: "The return type of an async method must be void, Task, Task<T>, a task-like type, IAsyncEnumerable<T>, or IAsyncEnumerator<T>"

File: README.md (hunk 3 above)
Comment (bartoval) [reply README.md]: Yes. I updated the title

**Review by thoraxe — CHANGES_REQUESTED** [review]
Ohhhh this is so close!

The last step is that we should be using the UID from the OAuth server (eg: the GitHub ID) for the player's UUID instead of asking them to input a player name.

**Comment by bartoval** [comment]
> Ohhhh this is so close!
> 
> The last step is that we should be using the UID from the OAuth server (eg: the GitHub ID) for the player's UUID instead of asking them to input a player name.

It seems to work.
ps:  there are some formatting changes because I activated this role in  .editorconfig "trim_trailing_whitespace = true".

**Comment by bartoval** [comment]
> It looks like you haven't completely removed the existing functionality of the loginscreen -- you just skip using it because the auth flow ends up calling `goToTheGame` which removes the loginscreen.
> 
> If authorization fails, it looks like nothing happens, and the login screen is still shown, which allows the user to enter their name and press the join button, which goes back to the old auth flow.
> 
> So I'm confused how this works.
> 
> Or is this because you are keeping the old flow to allow users to enter their own username when the auth is disabled?

The second one. 
I didn't touch the old functionality because it keeps the behaviour in case auth is disabled ( in dev mode).
We have a flag to enable/disable the auth as I mentionated in the Readme. If this flag is enabled we should use keycloak.

If auth fails we should see a page called not authorized and a retry button. I am talking about failure in the process of authentication.

File: Scenes/LoginScreen.cs [hunk 16]
```diff
@@ -26,21 +28,12 @@ private void _on_JoinButton_button_up()
...
-    }
-    else
-    {
-      // since we successfully joined the game, we can remove this node
-      // which is the login screen. removing the login screen "displays"
-      // the main game window
-      QueueFree();
-      MyGame.initializeGameUI();
-    }
+    MyGame.goToTheGame(textField.Text);
```
[the remark this answers is not in the document]
Comment (bartoval) [reply Scenes/LoginScreen.cs:31]: This is the case the auth is disabled. Basically I copied the the logic from this module and I created a function to call it in case we use the auth process or not

File: Scenes/MainScenes/Game.cs [hunk 17]
```diff
@@ -826,4 +844,46 @@ void deleteChildren(Node theNode)
...
+      loginScreen.GetNode<TextureRect>("AuthLoadingRect").Visible = false;
+
+      return;
+    }
+
+    loginScreen.GetNode<TextureRect>("NoAuthorizedRect").Visible = false;
+    loginScreen.GetNode<TextureRect>("AuthLoadingRect").Visible = false;
+
+    this.myUuid = authorization.getUserId();
+    this.goToTheGame(myUuid);
```
[the remark this answers is not in the document]
Comment (bartoval) [reply Scenes/MainScenes/Game.cs:861]: '_on_go_to_game(bool isAuthorized)'

This is happen when the auth process is completed.
Because the uuid is calculated from a token that I saved in the auth process, i created a function to calculate it there.

**Comment by bartoval** [comment]
> OK. I will probably merge this after I test it and then make some more changes. Thanks!!

I can remove the loginScreen functionality completely and pass a fake userId when auth is disabled .So the user can join dirctly.

**Comment by bartoval** [comment]
> OK. I will probably merge this after I test it and then make some more changes. Thanks!!

@thoraxe I made some changes and refactoring:

1) I tried to remove the old login page and join immediately in the game using a 'fake' userId.
However I found some syncronization issues when we call InitializeAMQP (that is an async method).
I thought of three ways to solve this problem but in the end I chose this one:
- The Game scene call the server and wait for a notification from it. 
- If the connection is ok go to the game (Maybe we can use an other PR to handle no happy paths)
It works with and without authentication enabled.

2) Moved the Auth logic in the LoginScreen because I think it is the responsible to handle this flow. (For example the Auth screens are here)

**Comment by bartoval** [comment]
@thoraxe I updated the PR and I changed the name of the prop **host** to **address** in the client.cfg

**Comment by bartoval** [comment]
@thoraxe Updated the documentation.
